### PR TITLE
Add warn-once deprecation warning to legacy sparse constructors

### PR DESCRIPTION
**Pull request by pearu**
Addresses https://github.com/pytorch/pytorch/issues/68323#issuecomment-1425174341

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94850

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer

### CODE REVIEW

**Review by amjames — APPROVED** [review]

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -553,21 +553,33 @@ Tensor legacy_sparse_tensor_generic_ctor_new(
...
+        "torch.sparse.SparseTensor(indices, values, *, device=) is deprecated."
+        "  Please use torch.sparse_coo_tensor(indices, values, dtype=, device=).");
     // Note: this signature doesn't have a dtype, even though it has a device;
     // it probably shouldn't have a device (we should infer it).
     auto deviceOptional = r.deviceOptional(2);
     check_legacy_ctor_device(dispatch_key, deviceOptional);
     at::OptionalDeviceGuard device_guard(deviceOptional);
     return at::sparse_coo_tensor(r.tensor(0), r.tensor(1));
   } else if (r.idx == 3) {
+    TORCH_WARN_ONCE(
```
Comment (cpuhrsch) [inline torch/csrc/utils/tensor_new.cpp:580]: You could try turning these into errors to see if they trigger CI failures? That could be one way to find hidden invocations deep within the stack that'd result in annoying warnings to users.

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Requesting changes so it doesn't get accidentally merged. We don't want these warnings to be triggered deep in the stack by other operations.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (pearu) [reply torch/csrc/utils/tensor_new.cpp:580]: @cpuhrsch 
```
grep -r -I --exclude "*~" 'sparse[.]\w*Tensor' test/
```
reveals 52 usages of `torch.sparse.SparseTensor` while in `torch/` there are none.
So, turning warn-once to an error will guarantee CI failures.

There exist legacy tests that test `SparseTensor` constructors. Do we want to remove these tests?

Comment (cpuhrsch) [reply torch/csrc/utils/tensor_new.cpp:580]: @pearu - yes if we're getting rid of these constructors let's also update those tests. Maybe we'll find a gap in functionality or uncover a bug.

File: test/test_sparse.py [hunk 2]
```diff
@@ -2531,11 +2526,11 @@ def test_sparse_add_coalesce(self, device, dtype):
         self.assertFalse(z._indices().numel() != 2 and z.is_coalesced())
 
     @onlyCUDA
-    def test_storage_not_null(self):
-        x = torch.cuda.sparse.FloatTensor(2)
+    def test_storage_not_null(self, device):
+        x = torch.empty(2, dtype=torch.float32, layout=torch.sparse_coo, device=device)
```
Comment (cpuhrsch) [inline test/test_sparse.py:2530]: Let's use the sparse constructors. I think this is supposed to go away @amjames?

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (pearu) [reply torch/csrc/utils/tensor_new.cpp]: After replacing warn-once with an error, CI still became green. So, no uncovered bugs here.

File: test/test_sparse.py (hunk 2 above)
Comment (amjames) [reply test/test_sparse.py:2530]: Yes, I am working on a change that will deprecate user facing `empty` and `empty_like` for sparse layouts. 

You should use the constructors directly, or `zeros` would be the closest replacement here.

Comment (pearu) [reply test/test_sparse.py:2530]: On second thought, we can also use `sparse_coo_tensor` constructor:
```python
x = torch.sparse_coo_tensor((2,), dtype=torch.float32)
```
which is the shortest variant among other options.

File: test/test_sparse.py [hunk 3]
```diff
@@ -82,6 +84,59 @@ def ignore_op(func):
             torch.ops.aten._values.default,
         )
 
+class TestSparse000(TestCase):
+    # we want to run the tests below once before all other test cases,
```
Comment (cpuhrsch) [inline test/test_sparse.py:88]: I don't think we can guarantee this order on some internal systems. I assume that's only important for development of this PR? If so can we change the name to something more expressive?

File: test/test_sparse.py [hunk 4]
```diff
@@ -82,6 +84,59 @@ def ignore_op(func):
...
+            x_ref = torch.sparse_coo_tensor((2, 3), dtype=torch.float64)
+            x = torch.sparse.DoubleTensor(2, 3)
+            self.assertEqual(x, x_ref)
+
+        for test_f in [f1, f2, f3, f4, f5]:
+            cerr = io.StringIO()
+            with redirect_stderr(cerr):
+                test_f()
+                test_f()  # checks warns once
+            FileCheck().check_count(str=test_f.__doc__, count=1, exactly=True).run(cerr.getvalue())
```
Comment (cpuhrsch) [inline test/test_sparse.py:138]: You might be able to use [`assertWarns`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarns).

File: test/test_sparse.py [hunk 5]
```diff
@@ -947,7 +996,7 @@ def test_not_in_place(x):
     def test_add_zeros(self, device, dtype, coalesced):
         def test_shape(sparse_dims, nnz, sizes):
             x, _, _ = self._gen_sparse(sparse_dims, nnz, sizes, dtype, device, coalesced)
-            zeros = torch.zeros(sizes, layout=torch.sparse_coo).to(x.device)
+            zeros = torch.sparse_coo_tensor(sizes).to(x.device)
```
Comment (cpuhrsch) [inline test/test_sparse.py:999]: nit: could pass device directly to constructor

File: test/test_sparse.py [hunk 6]
```diff
@@ -2821,54 +2849,33 @@ def test_tensor(indices, values, indices_equal, values_equal):
...
-        self.assertRaises(TypeError, lambda: torch.sparse.FloatTensor([6]))
-
     def test_legacy_new(self, device):
         i = torch.tensor([[0, 1, 1], [2, 0, 2]])
         v = torch.tensor([3., 4., 5.])
         size = torch.Size([2, 3])
         s = torch.sparse_coo_tensor(i, v, size)
 
         self.assertEqual(torch.sparse_coo, s.new(device='cpu').layout)
-        self.assertRaises(TypeError, lambda: s.new(v.storage()))
```
Comment (cpuhrsch) [inline test/test_sparse.py:2871]: Hm, why this change?

File: test/test_sparse.py [hunk 7]
```diff
@@ -2880,13 +2887,46 @@ def test_dtypes(self, device):
...
-        do_test_empty_full(self, all_sparse_dtypes, torch.sparse_coo, torch.device('cpu'))
+    @dtypes(*all_types_and_complex_and(torch.half, torch.bool, torch.bfloat16))
+    @parametrize('requires_grad', (True, False))
+    def test_empty_full(self, device, dtype, requires_grad):
+        if requires_grad and not (dtype.is_floating_point or dtype.is_complex):
+            self.skipTest(f'requires_grad==True requires float or complex dtype, got {dtype}')
+
+        self._test_empty_full(device, dtype, requires_grad)
         if torch.cuda.device_count() > 0:
-            do_test_empty_full(self, all_sparse_dtypes, torch.sparse_coo, None)
```
Comment (cpuhrsch) [inline test/test_sparse.py:2888]: Does this mean we can delete `do_test_empty_full`?

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Looking great! Just seen my inline comments.

File: test/test_sparse.py (hunk 3 above)
Comment (pearu) [reply test/test_sparse.py:88]: This test only works as intended when it is run before other tests that call legacy constructors because of the warn-once nature. That said, I now realize that other tests should never call the legacy constructors anyway, and if these do, these should be fixed. In fact, the test would be even stronger when the test is run after all other tests, although, I think we can consider it unnecessary.

In sum, I'll just rename the test class to TestSparseLegacyConstructors.

File: test/test_sparse.py (hunk 5 above)
Comment (pearu) [reply test/test_sparse.py:999]: right, fixed.

File: test/test_sparse.py (hunk 7 above)
Comment (pearu) [reply test/test_sparse.py:2888]: Not yet. `TestTensorCreation.test_empty_full` in test/test_tensor_creation_ops.py needs to be updated as well. It is slightly OT for this PR.

File: test/test_sparse.py (hunk 6 above)
Comment (pearu) [reply test/test_sparse.py:2871]: It resolves the following warning:
```
UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
```

File: test/test_sparse.py (hunk 4 above)
Comment (pearu) [reply test/test_sparse.py:138]: yes, thanks for the hint!

Btw, we have `assertWarnsOnceRegex` in torch/testing/_internal/common_utils.py that naming is misleading: it calls `set_warn_always(True)` to force `TORCH_WARN_ONCE` to become `TORCH_WARN` but that behavior is contradictory to the method name --- I would expect that it would implement a true test for the "warn once" feature.

File: test/test_sparse.py (hunk 3 above)
Comment (pearu) [reply test/test_sparse.py:88]: Fixed.

File: test/test_sparse.py (hunk 7 above)
Comment (cpuhrsch) [reply test/test_sparse.py:2888]: It seems safer to update `do_test_empty_full` with support for sparse layouts than creating another copy of it. Why not extend `do_test_empty_full`?

**Review by cpuhrsch — APPROVED** [review]
Looks good, see last comment about `do_test_empty_full`.

Thank you!

File: test/test_sparse.py (hunk 7 above)
Comment (pearu) [reply test/test_sparse.py:2888]: @cpuhrsch according to
https://github.com/pytorch/pytorch/blob/5f24b2b1f062e930c41b576e44195f9912dbc958/torch/testing/_internal/common_utils.py#L3676-L3677
the `do_test_empty_full` function will be dropped. So, this PR applies the FIXME to sparse layouts.
